### DISCUSSION TITLE
Looking for a working recipe to run osgearth with GL3 support on Linux

### DISCUSSION THREAD

**Discussion by elbarto1980** [pelicanmapping/osgearth#2389] [Support Q&A]
Hi,
I am starting this discussion hoping to have a place to share experiences with those of you that (like me) are working on Linux.
I have been using without any problem osgearth 3.2.0 and GL2 for a few years now and recently I have been trying, every now and then, to update my installation to the latest versions of OSG and osgearth, using GL3.

After several attempts, I have decided to sum up my linux _recipe_ in a Docker container that seems, in some extent, to works (unfortunately it has a few limitations).
Since the up-to-date vcpkg package for linux is still a [work in progress](https://github.com/microsoft/vcpkg/pull/34992), my dockerfile starts from scratch, installing packages on an Ubuntu 20.04 base image. The docker file is available in the pull request #2387 that I have opened.

The problem where I am stuck at the moment is related to model rendering: when I add a model, colors are mixed up... in a _deterministic_ way, it seems 😀... that I cannot fully understand 😦.
Here is what happens:
![285213391-5e1881f5-825a-4a8d-a0a9-d135bcba3e23](https://github.com/gwaldron/osgearth/assets/36400622/94c8b40e-2f41-4b8f-a3f2-4d40dc8ef954)

1. In the first (upper-left) image, the model has its own texture, but it has a red overlay that it takes from the color of the label;
2. Then I press the keyboard button 's' that shows statistics. As the "Frame Rate" text appears, the model's overlay color becomes white (upper-right image);
3. I press 's' again (lower-left image), and the overlay color turns to orange, as the GPU text in the statistics box;
4. I press 's' twice again (lower-right image) and the overlay is white, as the text in the "View" box.

In the terminal, it prints the message `Warning: Material::apply(State&) - not supported.`.

Something similar happens when I load a model that has materials, but no texture.
![285215791-ecd5ab62-8a94-4ca2-ba52-7220de133e26](https://github.com/gwaldron/osgearth/assets/36400622/5906b6cd-589d-4b4b-8c0b-0ea56057fc10)

This time, materials are not shown at all. The model only takes the color of the green label at first, and then the one of the statistics text. No warning messages are printed.

I will keep updating the Dockerfile if I have any updates, hoping to hearing from you as well. Consider the dockerfile as a sandbox.

Thank you!

**Comment by gwaldron** [comment]
The re-coloring is an example of "state leakage" and is typically caused by a model that leaves something undefined (like its colors array most commonly), in which case OSG just uses whatever "leaked" from the previous drawable. I would check that your 3D model has a color array defined.

The warning msg `Warning: Material::apply(State&) - not supported` comes from OSG, and will print when OSG sees a FFP state attribute (like Material) but can't apply it because FFP support is unavailable. It would be nice if you could suppress this OSG message, but there's no way to do so. The osgEarth ShaderGenerator is supposed to replace `osg::Material` with appropriate lighting uniforms in that case, but maybe it's not happening for some reason.

**Reply by elbarto1980** [reply]
Thank you, I will check the 3D model[s] I exported from Blender

**Comment by elbarto1980** [comment]
Hi, I have been able to fix the "state leakage" for the textured model by adding the options `--addMissingColors --fix-transparency-mode MAKE_ALL_STATESET_OPAQUE` to osgconv when converting the fbx model exported from Blender to an osgt file.
This solves the color issue but makes OSG print tons of `Warning: detected OpenGL error 'invalid enumerant' at after RenderBin::draw(..)` messages. They can only be silenced by setting `OSGNOTIFYLEVEL=FATAL`.

Now I have to address the Material issue for all the models that do not have a texture and I still have two `void StateSet::setGlobalDefaults() ShaderPipeline disabled.` messages (that should not [have consequences](https://github.com/gwaldron/osgearth/issues/2025#issuecomment-1095007685)) if I do not change the OSGNOTIFYLEVEL env variable.

I'll keep you posted...

**Reply by gwaldron** [reply]
The command line arg `--gldebug` (with osgearth_imgui or osgearth_viewer) may give you more information on the "invalid enumerant" error.
This is very likely OSG trying to execute a legacy FFP GL command.

**Reply by elbarto1980** [reply]
With `--gldebug`, osgearth_viewer prints the following messages:
```
[osgearth warning] GL (HIGH, 33350) -- GL_INVALID_ENUM error generated. Cannot enable <cap> in the current profile.
Warning: Material::apply(State&) - not supported.
[osgearth warning] GL (HIGH, 33350) -- GL_INVALID_ENUM error generated. Cannot enable <cap> in the current profile.
Warning: detected OpenGL error 'invalid enumerant' at after RenderBin::draw(..)
```

**Reply by gwaldron** [reply]
This means that something is called `glEnable(...)` with a mode that isn't supported in your OpenGL profile (GL3 or GLCORE).
One thing you can do is to set a breakpoint in the function `s_oe_gldebugproc` in `GLUtils.cpp`. The callstack should lead you to the offending code.
Also, can you please post the output of `osgearth_version --caps` ?

**Reply by elbarto1980** [reply]
I set the breakpoint (I am using LLDB as a debugger) but I cannot see the callstack. My copy of osgearth is built as RelWithDebInfo and it might not be enough (?).

Actually, I am starting to think that it is not the code (that is osgearth_viewer), but something **in** the model... because as soon as I move the model out of the view, it stops printing messages.

These are the caps output:
```
❯ ~/OSG3/osgearth/bin/osgearth_version --caps
[osgearth info] Hello, world.
[osgearth info] [Registry] Note: GDAL_DATA environment variable is not set
[osgearth info] [Capabilities] osgEarth Version:  3.5.0 build 149
[osgearth info] [Capabilities] GDAL Version:      3.0.4
[osgearth info] [Capabilities] OSG Version:       3.7.0
[osgearth info] [Capabilities] OSG GL3 Features:  yes
[osgearth info] [Capabilities] OSG FFP Available: no
[osgearth info] [Capabilities] CPU Cores:         16
[osgearth info] [Capabilities] GPU Renderer:      NVIDIA GeForce RTX 3070 Laptop GPU/PCIe/SSE2
[osgearth info] [Capabilities] GL Context/Driver: 4.6.0 NVIDIA 535.129.03 (460)
[osgearth info] [Capabilities] GL CORE Profile:   yes
```
I also uploaded the model [here](https://drive.google.com/file/d/1otvGqCuMt3DsZqecXUwzLZWR7YperfTO/view?usp=drive_link). This is how I added it to the earth file (same as in the _simple_model.earth_ example): 
```
  <Model name="Soldier" paged="true" max_range="5000">
      <url>Soldier.osgt.(0,0,3.14).trans.100.scale</url>
      <location lat="42.071589" long="12.332873"/>
  </Model>
```

**UPDATE:**
gdb provided a better insight:
```
[osgearth warning] GL (HIGH, 33350) -- GL_INVALID_ENUM error generated. Cannot enable <cap> in the current profile.
Warning: detected OpenGL error 'invalid enumerant' at after RenderBin::draw(..)

Thread 11 "osgearth_viewer" hit Breakpoint 1, (anonymous namespace)::s_oe_gldebugproc (source=33350, type=33356, id=<optimized out>, severity=37190, length=76, 
    message=0x7fffed0767a0 "GL_INVALID_ENUM error generated. Cannot enable <cap> in the current profile.", userParam=0x0)
    at /home/emanuele/gitrepo/Experimental/osgearth/src/osgEarth/GLUtils.cpp:419
419	            OE_WARN << "GL (" << s << ", " << source << ") -- " << message << std::endl;
(gdb) bt
#0  (anonymous namespace)::s_oe_gldebugproc (source=33350, type=33356, id=<optimized out>, severity=37190, length=76, 
    message=0x7fffed0767a0 "GL_INVALID_ENUM error generated. Cannot enable <cap> in the current profile.", userParam=0x0)
    at /home/emanuele/gitrepo/Experimental/osgearth/src/osgEarth/GLUtils.cpp:419
#1  0x00007fffeb65850c in ?? () from /lib/x86_64-linux-gnu/libnvidia-glcore.so.535.129.03
#2  0x00007fffeb65873b in ?? () from /lib/x86_64-linux-gnu/libnvidia-glcore.so.535.129.03
#3  0x00007ffff6dee73b in osg::State::apply(osg::StateSet const*) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosg.so.202
#4  0x00007ffff685771c in osgUtil::RenderLeaf::render(osg::RenderInfo&, osgUtil::RenderLeaf*) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#5  0x00007ffff6852028 in osgUtil::RenderBin::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#6  0x00007ffff685d083 in osgUtil::RenderStage::drawImplementation(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#7  0x00007ffff6852a9b in osgUtil::RenderBin::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#8  0x00007ffff685c424 in osgUtil::RenderStage::drawInner(osg::RenderInfo&, osgUtil::RenderLeaf*&, bool&) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#9  0x00007ffff685df1a in osgUtil::RenderStage::draw(osg::RenderInfo&, osgUtil::RenderLeaf*&) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#10 0x00007ffff6869616 in osgUtil::SceneView::draw() () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgUtil.so.202
#11 0x00007ffff663e33f in osgViewer::Renderer::draw() () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosgViewer.so.202
#12 0x00007ffff6d1eed1 in osg::GraphicsContext::runOperations() () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosg.so.202
#13 0x00007ffff6da0e7d in osg::OperationThread::run() () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosg.so.202
#14 0x00007ffff6d20fcf in non-virtual thunk to osg::GraphicsThread::run() () from /home/emanuele/OSG3/OpenSceneGraph/lib/libosg.so.202
#15 0x00007ffff444d9b3 in OpenThreads::ThreadPrivateActions::StartThread(void*) () from /home/emanuele/OSG3/OpenSceneGraph/lib/libOpenThreads.so.21
#16 0x00007ffff5c44609 in start_thread (arg=<optimized out>) at pthread_create.c:477
#17 0x00007ffff628e353 in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:95
```

**Reply by gwaldron** [reply]
Correct, it's some FFP state in the model that a GLCORE build will not accept.
Did you run your model through the shader generator?
```c++
osgEarth::ShaderGenerator gen;
gen.run(model);
```

**Reply by elbarto1980** [reply]
Yes, I did.
Actually, I get the same `Warning: detected OpenGL error 'invalid enumerant' at after RenderBin::draw(..)` message both:
- if I load the model **in the code** using `osgDB::readNodeFile`, and then run it through the shader generator as you suggested before;
- if I add the model **in the .earth file**, using the `<Model>` tag.

The models in the .earth file are run through the shader generator as well, aren't they?

**Reply by gwaldron** [reply]
OK.
So where are things now. If the model displaying correctly, but you just want to get rid of annoying messages?
Or are models with only materials (and no texture) displaying incorrectly?

**Reply by elbarto1980** [reply]
I think I can live with the annoying messages. At the moment, I am still figuring out how to render models with only materials, but I think I am close (a couple of days tops :crossed_fingers:).
I will let you know a.s.a.p.
Thank you again!

**Reply by elbarto1980** [reply]
Here I am with an update.
To render all the model, I have to use two different approaches for textured models and models that have just materials.

**Textured models:**
- must have everything defined, otherwise you can use the parameters `--addMissingColors --fix-transparency-mode MAKE_ALL_STATESET_OPAQUE` when converting them using `osgconv`;
- have to be run through the shader generator, otherwise I get a completely white model.

**osgParticles:**
- have to be run through the shader generator, otherwise I get big squares and rectangles.

**Models with materials (at least the ones I have):**
- have **not** to be run through the shader, neither loaded using the `.osgearth_shadergen` postfix, or I get the state leakage shown in previous posts.

The state leakage is probably caused by a lacking compatibility of the models I use with the osgearth shaders. I read the Blender 4.0 manuals regarding import-export for [Wavefront OBJ](https://docs.blender.org/manual/en/latest/files/import_export/obj.html#wavefront-obj) and [FBX](https://docs.blender.org/manual/en/latest/addons/import_export/scene_fbx.html), but tweaking the export settings does not really change the outcome.
I uploaded the osgt model with materials [here](https://drive.google.com/file/d/1ExCBirN7RK7BUTqTXfm_WsEAP2S-pCwS/view?usp=sharing) if anyone is interested in taking a peek under the hood...

**Reply by gwaldron** [reply]
Would it help if the shadergenerator were to add missing color arrays?

**Reply by elbarto1980** [reply]
To be honest, I have no clue... :roll_eyes:
